Src/CSharpier.VSCode/src/CSharpierProcessProvider.ts

```diff
@@ -0,0 +1,229 @@
+import { Disposable, TextEditor, window, workspace } from "vscode";
+import { LoggingService } from "./LoggingService";
+import * as path from "path";
+import * as semver from "semver";
+import { execSync } from "child_process";
+import { ICSharpierProcess, NullCSharpierProcess } from "./CSharpierProcess";
+import { CSharpierProcessSingleFile } from "./CSharpierProcessSingleFile";
+import { CSharpierProcessPipeMultipleFiles } from "./CSharpierProcessPipeMultipleFiles";
+import * as fs from "fs";
+import { InstallerService } from "./InstallerService";
+
+export class CSharpierProcessProvider implements Disposable {
+    warnedForOldVersion = false;
+    loggingService: LoggingService;
+    installerService: InstallerService;
+    csharpierPath: string;
+    warmingByDirectory: Record<string, boolean | undefined> = {};
+    csharpierVersionByDirectory: Record<string, string | undefined> = {};
+    csharpierProcessesByVersion: Record<string, ICSharpierProcess | undefined> = {};
+
+    constructor(loggingService: LoggingService) {
+        this.loggingService = loggingService;
+        this.installerService = new InstallerService(
+            this.loggingService,
+            this.killRunningProcesses,
+        );
+
+        let timeoutHandle: NodeJS.Timeout;
+        let setupKillRunningProcesses = () => {
+            // TODO we can't detect when the terminal gets focused
```

Comment:
The problem is that without us killing off csharpier, the user can't update csharpier at all. The command hangs and then eventually errors with the following. 
```
Failed to uninstall tool package 'csharpier': Access to the path
 'C:\Users\bela\.dotnet\tools\.store\csharpier\0.13.0' is denied.
```
There is no clear indication it is vscode that is keeping csharpier running. I figured out it was the extension because I know how the extension works, but the average user of csharpier would not have that knowledge.

My goal was to make the slow formatting invisible to the user, but I did find one edge case where a user will have a slow format document. https://github.com/belav/csharpier/issues/558

Reply:
I had a new thought laying in bed this morning! 

Make sure all the extensions use csharpier from a custom tool path, so that any user commands to install/upgrade/uninstall csharpier don't run into locked files and the extension doesn't need to continually kill and restart csharpier.
The extension would need to detect when the user changes versions. Or have an action to reload csharpier like you suggested. https://docs.microsoft.com/en-us/dotnet/core/tools/global-tools has details about the custom tool paths abilities.

Reply:
Right, I keep forgetting that windows doesn't allow updating executables in the background.
It's terrible that everyone has to keep doing workarounds like this because windows won't fix their shit, but oh well... :shrug: